src/main/scala/io/iohk/ethereum/db/storage/AppStateStorage.scala

```diff
@@ -2,27 +2,36 @@ package io.iohk.ethereum.db.storage
 
 import io.iohk.ethereum.db.dataSource.DataSource
 import io.iohk.ethereum.db.storage.AppStateStorage._
+import io.iohk.ethereum.db.storage.pruning.PruneResult
+import io.iohk.ethereum.db.storage.pruning.PruningMode.PruneFn
 
 /**
   * This class is used to store app state variables
   *   Key: see AppStateStorage.Keys
   *   Value: stored string value
   */
-class AppStateStorage(val dataSource: DataSource) extends KeyValueStorage[Key, Value, AppStateStorage]{
+class AppStateStorage(val dataSource: DataSource, pruneFn: PruneFn) extends KeyValueStorage[Key, Value, AppStateStorage]{
   type T = AppStateStorage
 
   val namespace: IndexedSeq[Byte] = Namespaces.AppStateNamespace
   def keySerializer: Key => IndexedSeq[Byte] = _.name.getBytes
   def valueSerializer: String => IndexedSeq[Byte] = _.getBytes
   def valueDeserializer: IndexedSeq[Byte] => String = (valueBytes: IndexedSeq[Byte]) => new String(valueBytes.toArray)
 
-  protected def apply(dataSource: DataSource): AppStateStorage = new AppStateStorage(dataSource)
+  protected def apply(dataSource: DataSource): AppStateStorage = new AppStateStorage(dataSource, pruneFn)
 
   def getBestBlockNumber(): BigInt =
     BigInt(get(Keys.BestBlockNumber).getOrElse("0"))
 
-  def putBestBlockNumber(bestBlockNumber: BigInt): AppStateStorage =
+  def putBestBlockNumber(bestBlockNumber: BigInt): AppStateStorage = {
+    // FIXME We need to decouple pruning from best block number storing in this fn
+    pruneFn(getLastPrunedBlock(), bestBlockNumber) match {
+      case PruneResult(_, 0) =>
```

Comment:
nvm, i think this will be a problem after `FastSync` so I'm going to change it as you suggested.